### PR TITLE
[Feature] Optimize forward metadata collection across dp ranks

### PR DESCRIPTION
**Pull request by jianzs**
### What this PR does / why we need it?

This PR introduces two optimizations for cases where data parallel size > 1:

1. Eliminates DP communication in `set_forward_context`
2. Implements HCCL for DP metadata communication, resulting in significant performance improvements for large DP configurations
   - Achieves ~20ms latency reduction with DP size of 64

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

CI passed.

### CODE REVIEW

**Comment by jianzs** [comment]
@NeverRaR PTAL

File: vllm_ascend/worker/model_runner_v1.py [hunk 1]
```diff
@@ -1775,6 +1779,7 @@ def _dummy_run(
         is_compile: bool = False,
         with_prefill: bool = True,
         skip_attn: bool = True,
+        num_tokens_across_dp: Optional[int] = None,
```
Comment (NeverRaR) [inline vllm_ascend/worker/model_runner_v1.py:1599]: Optional[int]  -> Optional[torch.Tensor]

**Comment by NeverRaR** [comment]
lgtm

**Comment by jianzs** [comment]
@Yikun @wangxiyuan  @ApsarasX  @ganyi1996ppo  ready to merge.

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -622,16 +620,16 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
...
-                        op=ReduceOp.MAX,
-                        group=get_dp_group().cpu_group)
-        return int(forward_metadata[0]), bool(forward_metadata[1] > 0)
+            self, num_tokens: int,
+            with_prefill: bool) -> tuple[torch.Tensor, bool]:
+        local_forward_metadata = torch.tensor([num_tokens, with_prefill],
+                                              device="npu",
+                                              dtype=torch.int32)
+        global_forward_metadata = get_dp_group().all_gather(
+            local_forward_metadata)
```
Comment (Angazenn) [inline vllm_ascend/worker/model_runner_v1.py]: Hello, I tried your PR and run offline inference with DP > 1, and I found that line 630 would throw a "dim out of range" error. It seems that you are using `all_gather` to get forward metadata across all dp ranks, but `local_forward_metadata` is actually an 1D tensor. Therefore, `global_forward_metadata` is also an 1D tensor. Perhaps codes should be changed like this:
```suggestion
        local_forward_metadata = torch.tensor([num_tokens, with_prefill],
                                              device="npu",
                                              dtype=torch.int32).unsqueeze(0)
        global_forward_metadata = get_dp_group().all_gather(
            local_forward_metadata, dim=0)
```

File: vllm_ascend/worker/worker_v1.py [hunk 3]
```diff
@@ -247,16 +247,30 @@ def pin_lora(self, lora_id: int) -> bool:
...
+            runner._get_forward_metadata_across_dp(-1, False)
+
         if runner.torchair_graph_enabled and not with_prefill:
-            max_num_tokens = runner.select_torchair_padded_batch_size(
+            max_num_tokens = int(num_tokens_across_dp.max().item())
+            num_tokens = runner.select_torchair_padded_batch_size(
                 max_num_tokens)
-        runner._dummy_run(max_num_tokens,
+        else:
+            num_tokens = 1
```
Comment (Angazenn) [inline vllm_ascend/worker/worker_v1.py:314]: why is it 1?

Comment (jianzs) [reply vllm_ascend/worker/worker_v1.py:314]: If graph mode is off, a dummy run only needs to be executed; computational requirements are not a factor.

File: vllm_ascend/worker/model_runner_v1.py (hunk 2 above)
Comment (jianzs) [reply vllm_ascend/worker/model_runner_v1.py]: Thanks. Indeed, this issue exists. Although all CI checks passed, it shows that our test cases didn't cover the DP scenarios 😂

**Comment by jianzs** [comment]
@Angazenn  PTAL

**Comment by jianzs** [comment]
@wangxiyuan  @ganyi1996ppo @Yikun  ready to merge.

**Comment by wangxiyuan** [comment]
torchair has made the code more and more complex and hard to maintain, I have a PR(https://github.com/vllm-project/vllm-ascend/pull/1661) to add torchair module, all torchair related code can be updated and changed there, I'll make the PR avaliable for review soon. Before that, I really don't want to merge anything about torchair change. Because it's very hard to review(I'm not sure if the change breaked anything else), Sorry.

**Comment by Yikun** [comment]
Wrong to submit the branch, pls feel free to open new PR.
